homeassistant/components/life360/helpers.py

```diff
@@ -1,7 +1,208 @@
+        errors["base"] = "invalid_auth"
+    except Life360Error as exc:
+        LOGGER.debug("Unexpected error communicating with Life360 server: %s", exc)
+        errors["base"] = "cannot_connect"
+
+    return authorization
+
+
+async def get_life360_data(
+    hass: HomeAssistant, api: Life360
+) -> dict[str, dict[str, Any]]:
+    """Fetch data from Life360."""
+
+    async def retrieve_data(
+        func: Callable[..., list[dict[str, Any]]], *args: Any
+    ) -> list[dict[str, Any]]:
+        try:
+            return await hass.async_add_executor_job(func, *args)
+        except LoginError as exc:
+            LOGGER.debug("Login error: %s", exc)
+            raise ConfigEntryAuthFailed from exc
+        except Life360Error as exc:
+            LOGGER.debug("%s: %s", exc.__class__.__name__, exc)
+            raise UpdateFailed from exc
+
+    data: dict[str, Any] = {"circles": {}, "members": {}}
+
+    for circle in await retrieve_data(api.get_circles):
+        circle_id = circle["id"]
+        circle_members = await retrieve_data(api.get_circle_members, circle_id)
+        circle_places = await retrieve_data(api.get_circle_places, circle_id)
+
+        data["circles"][circle_id] = {
+            "name": circle["name"],
+            "places": {
+                place["id"]: {"name": place["name"]}
+                | {
+                    k: float(v)
+                    for k, v in place.items()
+                    if k in ("latitude", "longitude", "radius")
+                }
+                for place in circle_places
+            },
+        }
+
+        for member in circle_members:
+            # Member isn't sharing location.
+            if not int(member["features"]["shareLocation"]):
+                continue
+
+            # Note that member may be in more than one circle. If that's the case just
+            # go ahead and process the newly retrieved data (overwriting the older
+            # data), since it might be slightly newer than what was retrieved while
+            # processing another circle.
+
+            first = member["firstName"]
+            last = member["lastName"]
+            if first and last:
+                name = " ".join([first, last])
+            else:
+                name = first or last
+
+            loc = member["location"]
+
+            place = loc["name"] or None
+
+            if place:
+                address: str | None = place
+            else:
+                address1 = loc["address1"] or None
+                address2 = loc["address2"] or None
+                if address1 and address2:
+                    address = ", ".join([address1, address2])
+                else:
+                    address = address1 or address2
+
+            speed = max(0, float(loc["speed"]) * SPEED_FACTOR_MPH)
+            if hass.config.units.is_metric:
+                speed = convert(speed, LENGTH_MILES, LENGTH_KILOMETERS)
+
+            data["members"][member["id"]] = {
```

Comment:
The member data should preferably be a dataclass too so we can keep it typed easily.